rewrite-test/src/main/kotlin/org/openrewrite/java/search/SemanticallyEqualTest.kt

```diff
@@ -0,0 +1,308 @@
+package org.openrewrite.java.search
+
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.Test
+import org.openrewrite.Formatting
+import org.openrewrite.Tree.randomId
+import org.openrewrite.java.JavaParser
+import org.openrewrite.java.tree.J
+import org.openrewrite.java.tree.JavaType
+
+interface SemanticallyEqualTest {
```

Comment:
One important mathematical attribute that equality typically has is the commutative property. This property holds that if `a == b`, then `b == a`. The concern I gave in another comment around some of the null checks being one-way was about violating this property. 

One way these tests could help to uphold the commutative property would be to always assert that both `SemanticallyEquals(a).visit(b)` and `SemanticallyEquals(b).visit(a)` should always return the same result.